File: backend/ideenmanagment-plattform/Authorization/Permissions.cs
namespace IdeaPlatform.Authorization;

/// <summary>Zentral definierte Berechtigungscodes (RBAC).</summary>
public static class Permissions
{
    public const string UsersRead = "users.read";
    public const string UsersManage = "users.manage";
    public const string RolesManage = "roles.manage";
    public const string CategoriesManage = "categories.manage";

    public const string IdeasCreate = "ideas.create";
    public const string IdeasRead = "ideas.read";
    public const string IdeasUpdateOwn = "ideas.update.own";
    public const string IdeasDeleteOwn = "ideas.delete.own";
    public const string IdeasDeleteAny = "ideas.delete.any";
    public const string IdeasModerate = "ideas.moderate";

    public const string CommentsCreate = "comments.create";
    public const string CommentsUpdateOwn = "comments.update.own";
    public const string CommentsDeleteOwn = "comments.delete.own";
    public const string CommentsDeleteAny = "comments.delete.any";

    public const string VotesCast = "votes.cast";
    public const string AttachmentsUpload = "attachments.upload";
    public const string AttachmentsDeleteAny = "attachments.delete.any";

    public const string NotificationsRead = "notifications.read";
    public const string AuditRead = "audit.read";

    public static IReadOnlyList<string> All { get; } = new[]
    {
        UsersRead, UsersManage, RolesManage, CategoriesManage,
        IdeasCreate, IdeasRead, IdeasUpdateOwn, IdeasDeleteOwn, IdeasDeleteAny, IdeasModerate,
        CommentsCreate, CommentsUpdateOwn, CommentsDeleteOwn, CommentsDeleteAny,
        VotesCast, AttachmentsUpload, AttachmentsDeleteAny,
        NotificationsRead, AuditRead
    };
}

public static class RoleNames
{
    public const string Mitarbeiter = "Mitarbeiter";
    public const string Moderator = "Moderator";
    public const string Administrator = "Administrator";
}

File: frontend/ideenmanagment-plattform/src/lib/permissions.ts
export const PERMISSIONS = {
  IdeasCreate: 'ideas.create',
  IdeasRead: 'ideas.read',
  IdeasUpdateOwn: 'ideas.update.own',
  IdeasDeleteOwn: 'ideas.delete.own',
  IdeasDeleteAny: 'ideas.delete.any',
  IdeasModerate: 'ideas.moderate',
  CommentsCreate: 'comments.create',
  CommentsUpdateOwn: 'comments.update.own',
  CommentsDeleteOwn: 'comments.delete.own',
  CommentsDeleteAny: 'comments.delete.any',
  VotesCast: 'votes.cast',
  AttachmentsUpload: 'attachments.upload',
  AttachmentsDeleteAny: 'attachments.delete.any',
  CategoriesManage: 'categories.manage',
  NotificationsRead: 'notifications.read',
  UsersRead: 'users.read',
  UsersManage: 'users.manage',
  RolesManage: 'roles.manage',
  AuditRead: 'audit.read',
} as const;

export type PermissionCode = (typeof PERMISSIONS)[keyof typeof PERMISSIONS];

export const ROLES = {
  Mitarbeiter: 'Mitarbeiter',
  Moderator: 'Moderator',
  Administrator: 'Administrator',
} as const;
export type RoleName = (typeof ROLES)[keyof typeof ROLES];

export function hasPermission(
  userPermissions: readonly string[] | undefined,
  required: string,
): boolean {
  if (!userPermissions) return false;
  return userPermissions.includes(required);
}

export function hasAnyPermission(
  userPermissions: readonly string[] | undefined,
  required: readonly string[],
): boolean {
  if (!userPermissions || required.length === 0) return false;
  return required.some((p) => userPermissions.includes(p));
}

export function hasAllPermissions(
  userPermissions: readonly string[] | undefined,
  required: readonly string[],
): boolean {
  if (!userPermissions) return false;
  return required.every((p) => userPermissions.includes(p));
}

export function hasRole(userRoles: readonly string[] | undefined, role: string): boolean {
  return !!userRoles?.includes(role);
}

